### PR TITLE
Initial testing of 4.16.2 vfs module

### PR DESCRIPTION
**Pull request by zefie**
Using Debian bookworm, which is currently Samba 4.16.2, I had to make the modifications below for it to work.

`fsp->fh->fd` was throwing compiler errors, `fsp_get_io_fd(fsp)` seems to be a suitable workaround.

For some reason, in the `renameat`, `linkat`, and `unlinkat` functions, `oldname->base_name` and `newname->base_name` were literally base names, and did not include paths within the share. My workaround fixes that as well.

Still in testing, but so far it seems stable.

Posting here for your thoughts.

### CODE REVIEW

File: samba-module/vfs_greyhole-samba-4.16.c [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+{
+	int result;
+	const struct loadparm_substitution *lp_sub = loadparm_s3_global_substitution();
+
+	result = SMB_VFS_NEXT_RENAMEAT(handle, srcfsp, oldname, dstfsp, newname);
+
+	if (result >= 0) {
+		gh_spoolf("rename\n%s\n%s/%s\n%s/%s\n\n",
+			lp_servicename(talloc_tos(), lp_sub, handle->conn->params->service),
+			srcfsp->fsp_name->base_name, oldname->base_name,
```
[the remark this answers is not in the document]
Comment (zefie) [reply samba-module/vfs_greyhole-samba-4.16.c]: I didn't think to check 4.15's code. I started over from v4.15's code and added the `fsp->fh->fd` to `fsp_get_io_fd(fsp)` conversion. I also took the final v4.16 and copied it to 4.x, then squashed all the commits.

I also enabled gpg signing for the commit.

[the remark this answers is not in the document]
Comment (zefie) [reply samba-module/vfs_greyhole-samba-4.16.c]: Also I was wondering if `fsp_get_io_fd` is appropriate, or if `fsp_get_pathref_fd` would be more appropriate. Both seem to work equally in the testing I have done.

[the remark this answers is not in the document]
Comment (gboudreau) [reply samba-module/vfs_greyhole-samba-4.16.c]: I think per https://wiki.samba.org/index.php/The_New_VFS#When_to_use_fsp_get_io_fd.28.29_or_fsp_get_pathref_fd.28.29, we should use `fsp_get_pathref_fd` : 
> if you're doing a logical comparison of fd values, use fsp_get_pathref_fd().

[the remark this answers is not in the document]
Comment (zefie) [reply samba-module/vfs_greyhole-samba-4.16.c]: Alright, I switched it over to `fsp_get_pathref_fd`, seems to work well.

[the remark this answers is not in the document]
Comment (gboudreau) [reply samba-module/vfs_greyhole-samba-4.16.c]: OK, let me know when you're done testing, and are satisfied that everything seems to work as expected.  
Of note: I have a test script I use to test big changes : `test_cases.php` in the root of the repo.
If you change lines 26-40, you should be able to use it.

- `log_file` should be `/var/log/greyhole.log`, unless you changed the `log_file` in greyhole.conf
- `test_dir` should be the path to the locally mounted share (should be empty)
- `share_dir` should be the LZ of that share
- `pool_dirs` should be path to that share, in all your storage pool drives

[the remark this answers is not in the document]
Comment (zefie) [reply samba-module/vfs_greyhole-samba-4.16.c]: Have had no problems, and been actively using it. Gonna mark this resolved.

**Comment by gboudreau** [comment]
Thanks for your help on this. I'll compile and package the VFS module for 4.16 from this.

**Pull request merged**